extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/LifecycleEventsBuildStep.java

```diff
@@ -4,14 +4,23 @@
 
 import java.util.List;
 
+import javax.enterprise.context.BeforeDestroyed;
+
 import io.quarkus.arc.runtime.ArcDeploymentTemplate;
+import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
 import io.quarkus.deployment.annotations.Record;
 import io.quarkus.deployment.builditem.ServiceStartBuildItem;
 import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
+import io.quarkus.deployment.builditem.substrate.ReflectiveClassBuildItem;
 
 public class LifecycleEventsBuildStep {
 
+    @BuildStep
+    public void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass) {
+        reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, BeforeDestroyed.class.getName()));
```

Comment:
Why is this actually needed? Why only `BeforeDestroyed` and not other lifecycle qualifiers like `Initialized`?

Reply:
Without it, `members` is empty here: https://github.com/quarkusio/quarkus/blob/master/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/Qualifiers.java#L47 

I added it for `BeforeDestroyed` because it was causing problems for Kafka: https://github.com/smallrye/smallrye-reactive-messaging/blob/master/smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/KafkaMessagingProvider.java#L33 

Closing Request scope was causing the `terminate` method to be executed.

As for the other lifecycle qualifiers, I think you are right that they should be added here too. What are they?

Reply:
Hm, this probably reveals a more general issue - all built-in qualifiers should be probably registered for reflection. In fact, `native-image` should register reflective metadata automatically for all qualifiers for which we have generated annotation literals (at least this was the default behavior until recently). And there should be one for `@BeforeDestroyed` used in the generated observer class for `KafkaMessagingProvider#terminate()`. 

@michalszynkiewicz anyway, `KafkaMessagingProvider#terminate()` should be invoked when the app is shut down as it observes `@BeforeDestroyed(ApplicationScoped.class)`.

Reply:
Agree on the first paragraph, don't understand the second.
Without the change `KafkaMessagingProvider#terminate()` gets called on any `@BeforeDestroyed`, with it it gets called on the app shut down.

Reply:
Np, I misread your comment ;-)